src/dstore/pmix_esh.c

```diff
@@ -0,0 +1,1374 @@
+    tmp_dir = strdup(pmix_tmp_directory());
+    _set_constants_from_env();
+    max_ns_num = (initial_segment_size - sizeof(size_t) - sizeof(int))/sizeof(ns_seg_info_t);
+    max_meta_elems = (meta_segment_size - sizeof(size_t))/sizeof(rank_meta_info);
+
+    /* create a lock file to prevent clients from reading while server is writing to the shared memory.
+     * This situation is quite often, especially in case of direct modex when clients might ask for data
+     * simultaneously.*/
+    lockfile_name = malloc(256);
+    snprintf(lockfile_name, 256, "%s/dstore_sm.lock", tmp_dir);
+    if (_is_server()) {
+        lockfd = open(lockfile_name, O_CREAT | O_RDWR | O_EXCL, 0600);
+        /* if previous launch was crashed, the lockfile might not be deleted and unlocked,
+         * so we delete it and create a new one. */
+        if (-1 == lockfd) {
+            unlink(lockfile_name);
+            lockfd = open(lockfile_name, O_CREAT | O_RDWR, 0600);
+        }
+    } else {
+        lockfd = open(lockfile_name, O_RDWR);
+    }
+    if (-1 == lockfd) {
+        PMIX_ERROR_LOG(PMIX_ERROR);
+        return PMIX_ERROR;
+    }
+
+    if (_is_server()) {
+        seg = _create_new_segment(INITIAL_SEGMENT, NULL, 0);
+    } else {
+        seg = _attach_new_segment(INITIAL_SEGMENT, NULL, 0);
+    }
+    if (NULL != seg) {
+        global_sm_seg_first = seg;
+        global_sm_seg_last = global_sm_seg_first;
+        return PMIX_SUCCESS;
+    }
+    return PMIX_ERROR;
+}
+
+int _esh_finalize(void)
+{
+    PMIX_OUTPUT_VERBOSE((10, pmix_globals.debug_output,
+                         "%s:%d:%s", __FILE__, __LINE__, __func__));
+
+    _delete_sm_desc(global_sm_seg_first);
+    PMIX_LIST_DESTRUCT(&namespace_info_list);
+    if (NULL != tmp_dir) {
+        free(tmp_dir);
+    }
+    close(lockfd);
+    if (_is_server()) {
+        unlink(lockfile_name);
+    }
+    free(lockfile_name);
+
+    pmix_sm_finalize();
+
+    return PMIX_SUCCESS;
+}
+
+
+int _esh_store(const char *nspace, int rank, pmix_kval_t *kv)
+{
+    int rc;
+    ns_track_elem_t *elem;
+    pmix_buffer_t pbkt, xfer;
+    ns_seg_info_t ns_info;
+
+    if (NULL == kv) {
+        return PMIX_ERROR;
+    }
+
+    PMIX_OUTPUT_VERBOSE((10, pmix_globals.debug_output,
+                         "%s:%d:%s: for %s:%d", __FILE__, __LINE__, __func__, nspace, rank));
+
+    /* set exclusive lock */
+    flock(lockfd, LOCK_EX);
+
+    /* First of all, we go through local track list (list of ns_track_elem_t structures)
+     * and look for an element for the target namespace.
+     * If it is there, then shared memory segments for it are created, so we take it.
+     * Otherwise, create a new element, fill its fields, create corresponding meta
+     * and data segments for this namespace, add it to the local track list,
+     * and put this info (ns_seg_info_t) to the initial segment. If initial segment
+     * if full, then extend it by creating a new one and mark previous one as full.
```

Comment:
Sounds like this coment is outdated or is untrue.
The only thing that `_get_track_elem_for_namespace()` does is returning existing list element or creating the new one. No initialization or shmem segments are created. This is done below here in this function.